libs/langchain-mongodb/langchain_mongodb/retrievers/self_querying.py

```diff
@@ -0,0 +1,217 @@
+from langchain_core.language_models import BaseLanguageModel
+from langchain_core.runnables import Runnable
+from langchain_core.structured_query import (
+    Comparator,
+    Comparison,
+    Operation,
+    Operator,
+    StructuredQuery,
+    Visitor,
+)
+from langchain_core.vectorstores import VectorStore
+from pydantic import Field
+
+from langchain_mongodb import MongoDBAtlasVectorSearch
+
+
+class MongoDBStructuredQueryTranslator(Visitor):
+    """Translator between MongoDB Query API and LangChain's StructuredQuery.
+
+    With Vector Search Indexes, one can index boolean, date, number, objectId, string,
+    and UUID fields to pre-filter your data.
+    Filtering your data is useful to narrow the scope of your semantic search
+    and ensure that not all vectors are considered for comparison.
+    It reduces the number of documents against which to run similarity comparisons,
+    which can decrease query latency and increase the accuracy of search results.
+    """
+
+    """Available logical comparators."""
+    allowed_comparators = [
+        Comparator.EQ,
+        Comparator.NE,
+        Comparator.GT,
+        Comparator.GTE,
+        Comparator.LT,
+        Comparator.LTE,
+        Comparator.IN,
+        Comparator.NIN,
+    ]
+
+    """Available logical operators."""
+    allowed_operators = [Operator.AND, Operator.OR]
+
+    ## Convert an operator or a comparator to Mongo Query Format
+    def _format_func(self, func: Union[Operator, Comparator]) -> str:
+        self._validate_func(func)
+        map_dict = {
+            Operator.AND: "$and",
+            Operator.OR: "$or",
+            Comparator.EQ: "$eq",
+            Comparator.NE: "$ne",
+            Comparator.GTE: "$gte",
+            Comparator.LTE: "$lte",
+            Comparator.LT: "$lt",
+            Comparator.GT: "$gt",
+            Comparator.IN: "$in",
+            Comparator.NIN: "$nin",
+        }
+        return map_dict[func]
+
+    def visit_operation(self, operation: Operation) -> Dict:
+        args = [arg.accept(self) for arg in operation.arguments]
+        return {self._format_func(operation.operator): args}
+
+    def visit_comparison(self, comparison: Comparison) -> Dict:
+        if comparison.comparator in [Comparator.IN, Comparator.NIN] and not isinstance(
+            comparison.value, list
+        ):
+            comparison.value = [comparison.value]
+        comparator = self._format_func(comparison.comparator)
+        attribute = comparison.attribute
+        return {attribute: {comparator: comparison.value}}
+
+    def visit_structured_query(
+        self, structured_query: StructuredQuery
+    ) -> Tuple[str, dict]:
+        if structured_query.filter is None:
+            kwargs = {}
+        else:
+            kwargs = {"pre_filter": structured_query.filter.accept(self)}
+        return structured_query.query, kwargs
+
+
+class MongoDBAtlasSelfQueryRetriever(SelfQueryRetriever):
```

Comment:
This is the class that needs the `close()` method, to align with our other classes.

Reply:
> This is the class that needs the `close()` method, to align with our other classes.

Ah. I was working late and multitasking. That makes sense. Done.

https://evergreen.mongodb.com/version/67dc3f34e942f7000767aef6?redirect_spruce_users=true